Faster lub

jvilk-stripe opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->

A continuation of https://github.com/sorbet/sorbet/pull/3241 which makes lub much faster for large `OrType`s.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

Sorbet is slow to typecheck files that `case` over a large enum with hundreds of members. That slowness stems from lub.

This change results in a large speedup on these files (10X+).

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

jvilk-stripe commented on core/types/subtyping.cc:
```diff
@@ -221,7 +258,9 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
-        bool changed = false;
+        bool changedFromT2 = false;
+        // If klasses are equal, then it's possible that t1s <: t2s.
+        bool changedFromT1 = a1->klass != a2->klass;
```
This is the correctness fix.

Before this change:
* `lub(T::Array[T.untyped], T::Array[X])` would return a `T::Array[T.untyped]` that is pointer equivalent with the first parameter.
* `lub(T::Array[X], T::Array[T.untyped])` would return a fresh `T::Array[T.untyped]` that is not pointer equivalent with the second parameter.

After this change:
* Both `lub(T::Array[T.untyped], T::Array[X])` and `lub(T::Array[X], T::Array[T.untyped])` return the same `T::Array[T.untyped]` pointer passed in to lub.

jvilk-stripe commented:
This PR exposes new errors on Stripe's codebase; some look legitimate, others look false.

So I should probably complete that investigation prior to a full review.

jvilk-stripe commented:
This PR is now ready for review.

elliottt commented on test/testdata/infer/lub_and_glb_corner_cases.rb:
```diff
@@ -51,4 +51,27 @@ def generic_class_glb_collapse
+    end
+    
+    # T.any(nil, T::Array[T.nilable(String)], T::Array[T.untyped]) => T.nlable(T::Array[T.untyped])
+    T.assert_type!(xs, T.nilable(T::Array[T.untyped]))
```
Could you use `T.reveal_type` here instead?

jvilk-stripe replied:
Talked in DM; `reveal_type` shows the actual type Sorbet sees and doesn't reduce it, so it's better for testing lub. `T.assert_type!` calls into lub and may reduce the type further.

elliottt commented on test/testdata/infer/lub_and_glb_corner_cases.rb:
```diff
@@ -51,4 +51,27 @@ def generic_class_glb_collapse
+  def generic_class_nilable_array_tuple_lub
+    a = T.let(nil, T.nilable(T::Array[Float]))
+    b = T.let(nil, T.nilable(T::Array[Integer]))
+    T.assert_type!([a, b], [T.nilable(T::Array[Float]), T.nilable(T::Array[Integer])])
```
Same for this one.

elliottt commented on test/testdata/infer/fast_big_lubs.rb:
```diff
@@ -0,0 +1,533 @@
+# typed: true
```
Could you move this file into `test/testdata/perf`? We have a few other performance regression tests there already.

jvilk-stripe replied:
Good idea; can do

elliottt reviewed (approved):
The change to `Types::lub` makes sense to me, thanks for the additional test cases! Were there any other changes to @DarkDimius's original PR that weren't obvious? It looked like there were some small changes to `lubDistributeOr`, but it wasn't clear if the ended up affecting its behavior.

I'd also be interested in seeing a build with the new errors raised, when you get a chance 👍

jvilk-stripe commented:
Links to internal Stripe builds:

* With typechecking errors: https://cibot.corp.stripe.com/builds/bui_Hx3kQn7Zor4oS3
* Clean build with fixes from a branch: https://cibot.corp.stripe.com/builds/bui_HzaU9panlyfxC9

jvilk-stripe commented:
>Were there any other changes to @DarkDimius's original PR that weren't obvious? It looked like there were some small changes to `lubDistributeOr`

Nope. See: https://github.com/sorbet/sorbet/compare/98939cc..f9974ea

I added some comments, made an `if` into an `else if`, and fixed a bug in the metric Dmitry's original code was reporting.
